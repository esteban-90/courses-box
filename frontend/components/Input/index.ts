export * from './Input'
export * from './Feedback'
