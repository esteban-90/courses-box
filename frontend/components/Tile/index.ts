export * from './Tile'
export * from './CenteredTile'
