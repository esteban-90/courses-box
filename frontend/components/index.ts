export * from './Button'
export * from './Checkbox'
export * from './Course'
export * from './Icon'
export * from './IconButton'
export * from './Input'
export * from './Layout'
export * from './Link'
export * from './Logo'
export * from './Switch'
export * from './Tile'
