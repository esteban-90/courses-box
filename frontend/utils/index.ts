import qs from 'qs'
import { apiUrl } from '@/config'
import type { CoursePayload, CoursesPayload, ErrorPayload, Course as CourseType } from '@/types'

const getData = async (param: string) => {
  const response = await fetch(`${apiUrl}/courses${param}`, { method: 'GET' })
  const result = await response.json()
  const { status } = response
  const error = status >= 400 ? (result as ErrorPayload) : undefined
  const success = status >= 200 && status < 400 ? (result as CoursesPayload | CoursePayload) : undefined

  return { data: success?.data, error: error?.error?.message ?? '' }
}

type Data<D> = Promise<{ data: D; error: string }>

export const getCourses = () => getData('?populate=*') as Data<CourseType[]>
export const getCourseById = (id: string) => getData(`/${id}?populate=*`) as Data<CourseType>

export const getCoursesByQuery = (q: string) => {
  const query = qs.stringify(
    {
      populate: '*',
      filters: {
        $or: [{ title: { $containsi: q } }, { subtitle: { $containsi: q } }, { description: { $containsi: q } }],
      },
    },
    { encodeValuesOnly: true }
  )

  return getData(`?${query}`) as Data<CourseType[]>
}
