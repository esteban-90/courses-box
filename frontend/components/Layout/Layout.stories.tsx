import { ComponentStoryObj, ComponentMeta } from '@storybook/react'
import { expect } from '@storybook/jest'
import { screen } from '@storybook/testing-library'
import { Layout } from '@/components/Layout'

export default {
  title: 'Content/Layout',
  component: Layout,
} as ComponentMeta<typeof Layout>

const child = (
  <>
    <h1>Main article area</h1>
    <p>
      In this layout, we display the areas in source order for any screen less than 500 pixels wide. We go to a two
      column layout, and then to a three column layout by redefining the grid,and the placement of items on the grid.
    </p>
  </>
)

export const BasicLayout: ComponentStoryObj<typeof Layout> = {
  async play() {
    await expect(screen.getByText('Main article area')).toBeInTheDocument()
  },

  args: {
    children: child,
  },
}
