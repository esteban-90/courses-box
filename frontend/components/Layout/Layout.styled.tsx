import styled from '@emotion/styled'
import { Input as _Input } from '@/components/Input'
import { Link } from '@/components/Link'
import { Logo as _Logo } from '@/components/Logo'

export const Content = styled.main`
  grid-area: content;
  min-height: 84vh;
  margin-top: 1rem;
`

export const Footer = styled.footer`
  grid-area: footer;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
`

export const Input = styled(_Input)`
  grid-area: search;
  width: 100%;
  height: 3.5rem;
`

export const Logo = styled(_Logo)`
  grid-area: header;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  height: 3.5rem;

  & .logo_full {
    display: none;
  }

  @media (min-width: 560px) {
    & .logo_short {
      display: none;
    }

    & .logo_full {
      display: inline;
    }
  }
`

export const LogoLink = styled(Link)`
  padding-right: 1vw;
`

export const Navigation = styled.nav`
  grid-area: nav;
  display: flex;
  justify-content: space-around;
  align-items: center;
  margin: 0 2vmin;
`

export const Wrapper = styled.div`
  display: grid;
  gap: 0.1rem;
  padding: 0.5rem;
  color: ${({ theme }) => theme.font.regular};
  background-color: ${({ theme }) => theme.background};

  grid-template-areas:
    'header nav'
    'search search'
    'content content'
    'footer footer';

  nav {
    flex-direction: row;
    justify-content: flex-end;
    gap: 5vmin;
  }

  @media (min-width: 500px) {
    grid-template-columns: 1fr 3fr;
  }

  @media (min-width: 960px) {
    grid-template-columns: 1fr 4fr 2fr;

    grid-template-areas:
      'header search nav'
      'content content content'
      'footer footer footer';
  }
`
