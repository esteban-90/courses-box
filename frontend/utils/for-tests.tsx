import { FC, ReactElement, ReactNode } from 'react'
import { Provider as StoreProvider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import { render, RenderResult } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'
import { Layout } from '@/components'
import { rootReducer } from '@/store'
import { Themes } from '@/styles/themes'
import { StoreAndRenderOptions } from '@/types'

const customRender = (ui: ReactElement, { preloadedState, ...options }: StoreAndRenderOptions = {}): RenderResult => {
  const store = configureStore({ reducer: rootReducer, preloadedState })

  const Wrapper: FC<{ children: ReactNode }> = ({ children }): JSX.Element => {
    return (
      <StoreProvider store={store}>
        <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
      </StoreProvider>
    )
  }

  return render(ui, { wrapper: Wrapper, ...options })
}

const pageRender = (ui: ReactElement, { preloadedState, ...options }: StoreAndRenderOptions = {}): RenderResult => {
  const store = configureStore({ reducer: rootReducer, preloadedState })

  const Wrapper: FC<{ children: ReactNode }> = ({ children }): JSX.Element => {
    return (
      <StoreProvider store={store}>
        <Layout>{children}</Layout>
      </StoreProvider>
    )
  }

  return render(ui, { wrapper: Wrapper, ...options })
}

export * from '@testing-library/react'
export { customRender as render, pageRender }
