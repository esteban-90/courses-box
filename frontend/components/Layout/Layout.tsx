import { FC, useState, useEffect, useLayoutEffect } from 'react'
import { ThemeProvider } from '@emotion/react'
import NextLink from 'next/link'
import { IconButton } from '@/components/IconButton'
import { Link } from '@/components/Link'
import { Themes } from '@/styles/themes'
import { LayoutProps } from '@/types'
import * as Styled from './Layout.styled'

const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayoutEffect : useEffect

export const Layout: FC<LayoutProps> = ({ children }): JSX.Element => {
  const [isDark, setIsDark] = useState(true)
  const theme = Themes[isDark ? 'dark' : 'light']

  const toggleTheme = () => {
    localStorage.setItem('courses-box-theme', isDark ? 'light' : 'dark')
    setIsDark(!isDark)
  }

  useIsomorphicLayoutEffect(() => {
    const savedTheme = localStorage.getItem('courses-box-theme')
    const savedThemeExists = savedTheme !== null
    const savedThemeIsDark = savedTheme === 'dark'
    const preferenceIsDark = window.matchMedia('prefers-color-scheme: dark').matches
    const themeIsDark = savedThemeExists ? savedThemeIsDark : preferenceIsDark

    setIsDark(themeIsDark)
  }, [])

  return (
    <ThemeProvider theme={theme}>
      <Styled.Wrapper>
        <NextLink href='/' passHref>
          <Styled.LogoLink>
            <Styled.Logo size={3}>
              <span className='logo_short'>C8X</span>
              <span className='logo_full'>CoursesBox</span>
            </Styled.Logo>
          </Styled.LogoLink>
        </NextLink>

        <Styled.Navigation>
          <NextLink href='/all' passHref>
            <Link>All</Link>
          </NextLink>

          <NextLink href='/login' passHref>
            <IconButton name='Login' size={1} />
          </NextLink>

          <IconButton name={isDark ? 'Sun' : 'Moon'} size={1} onClick={toggleTheme} />
        </Styled.Navigation>

        <Styled.Input icon='Search' placeholder='Search' onChange={() => {}} />
        <Styled.Content>{children}</Styled.Content>

        <Styled.Footer>
          <p>
            &copy; {new Date().getFullYear()}&nbsp;
            <Link href='https://github.com/esteban-90' target='_blank'>
              Esteban V.M.
            </Link>
            &nbsp;
          </p>
          <p>All rights reserved</p>
        </Styled.Footer>
      </Styled.Wrapper>
    </ThemeProvider>
  )
}
