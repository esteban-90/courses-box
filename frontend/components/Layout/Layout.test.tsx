import userEvent from '@testing-library/user-event'
import { Layout } from '@/components/Layout'
import { render, screen, act } from '@/utils'

describe('Layout test cases', () => {
  const layoutContent = (
    <>
      <h1>Main article area</h1>
      <p>
        In this layout, we display the areas in source order for any screen less than 500 pixels wide. We go to a two
        column layout, and then to a three column layout by redefining the grid,and the placement of items on the grid.
      </p>
    </>
  )

  const layoutElement = <Layout>{layoutContent}</Layout>

  it('Render check', () => {
    const { asFragment } = render(layoutElement)
    expect(asFragment()).toMatchSnapshot()
  })

  it('Theme toggle check', async () => {
    localStorage.setItem('courses-box-theme', 'light')
    void (window.matchMedia as jest.Mock).mockReturnValue({ matches: true })
    render(layoutElement)

    const themeToggler = screen.getByRole('button', { name: 'Moon' })
    expect(themeToggler).toBeInTheDocument()

    await act(async () => await userEvent.click(themeToggler))
    expect(localStorage.getItem('courses-box-theme')).toBe('dark')
    expect(screen.getByRole('button', { name: 'Sun' })).toBeInTheDocument()
  })
})
