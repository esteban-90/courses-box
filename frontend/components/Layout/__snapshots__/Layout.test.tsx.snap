// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout test cases: should check render 1`] = `
<DocumentFragment>
  <div
    class="css-19wldrt"
  >
    <a
      class="css-1nhvv70"
      href="/"
    >
      <header
        class="css-bg0f53"
        size="3"
      >
        <span
          class="logo_short"
        >
          C8X
        </span>
        <span
          class="logo_full"
        >
          CoursesBox
        </span>
      </header>
    </a>
    <nav
      class="css-1sstk05"
    >
      <button
        class="css-1aqtdpz"
        size="2"
        title="UserAdd"
      >
        <svg
          aria-label="UserAdd"
          class="css-z3g3v6"
          fill="currentColor"
          height="1rem"
          href="/register"
          role="img"
          viewBox="0 0 1024 1024"
          width="1rem"
        >
          <path
            d="M678.3 642.4c24.2-13 51.9-20.4 81.4-20.4h.1c3 0 4.4-3.6 2.2-5.6a371.67 371.67 0 00-103.7-65.8c-.4-.2-.8-.3-1.2-.5C719.2 505 759.6 431.7 759.6 349c0-137-110.8-248-247.5-248S264.7 212 264.7 349c0 82.7 40.4 156 102.6 201.1-.4.2-.8.3-1.2.5-44.7 18.9-84.8 46-119.3 80.6a373.42 373.42 0 00-80.4 119.5A373.6 373.6 0 00137 888.8a8 8 0 008 8.2h59.9c4.3 0 7.9-3.5 8-7.8 2-77.2 32.9-149.5 87.6-204.3C357 628.2 432.2 597 512.2 597c56.7 0 111.1 15.7 158 45.1a8.1 8.1 0 008.1.3zM512.2 521c-45.8 0-88.9-17.9-121.4-50.4A171.2 171.2 0 01340.5 349c0-45.9 17.9-89.1 50.3-121.6S466.3 177 512.2 177s88.9 17.9 121.4 50.4A171.2 171.2 0 01683.9 349c0 45.9-17.9 89.1-50.3 121.6C601.1 503.1 558 521 512.2 521zM880 759h-84v-84c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v84h-84c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h84v84c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-84h84c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </button>
      <button
        class="css-1aqtdpz"
        size="2"
        title="Login"
      >
        <svg
          aria-label="Login"
          class="css-z3g3v6"
          fill="currentColor"
          height="1rem"
          href="/login"
          role="img"
          viewBox="0 0 1024 1024"
          width="1rem"
        >
          <defs>
            <style />
          </defs>
          <path
            d="M521.7 82c-152.5-.4-286.7 78.5-363.4 197.7-3.4 5.3.4 12.3 6.7 12.3h70.3c4.8 0 9.3-2.1 12.3-5.8 7-8.5 14.5-16.7 22.4-24.5 32.6-32.5 70.5-58.1 112.7-75.9 43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 32.6 32.5 58.1 70.4 76 112.5C865.7 417.8 875 464.1 875 512c0 47.9-9.4 94.2-27.8 137.8-17.8 42.1-43.4 80-76 112.5s-70.5 58.1-112.7 75.9A352.8 352.8 0 01520.6 866c-47.9 0-94.3-9.4-137.9-27.8A353.84 353.84 0 01270 762.3c-7.9-7.9-15.3-16.1-22.4-24.5-3-3.7-7.6-5.8-12.3-5.8H165c-6.3 0-10.2 7-6.7 12.3C234.9 863.2 368.5 942 520.6 942c236.2 0 428-190.1 430.4-425.6C953.4 277.1 761.3 82.6 521.7 82zM395.02 624v-76h-314c-4.4 0-8-3.6-8-8v-56c0-4.4 3.6-8 8-8h314v-76c0-6.7 7.8-10.5 13-6.3l141.9 112a8 8 0 010 12.6l-141.9 112c-5.2 4.1-13 .4-13-6.3z"
          />
        </svg>
      </button>
      <button
        class="css-1aqtdpz"
        size="2"
        title="Moon"
      >
        <svg
          aria-label="Moon"
          class="css-z3g3v6"
          fill="currentColor"
          height="1rem"
          role="img"
          viewBox="0 0 16 16"
          width="1rem"
        >
          <path
            d="M6 .278a.768.768 0 01.08.858 7.208 7.208 0 00-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 01.81.316.733.733 0 01-.031.893A8.349 8.349 0 018.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 016 .278zM4.858 1.311A7.269 7.269 0 001.025 7.71c0 4.02 3.279 7.276 7.319 7.276a7.316 7.316 0 005.205-2.162c-.337.042-.68.063-1.029.063-4.61 0-8.343-3.714-8.343-8.29 0-1.167.242-2.278.681-3.286z"
          />
          <path
            d="M10.794 3.148a.217.217 0 01.412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 010 .412l-1.162.387a1.734 1.734 0 00-1.097 1.097l-.387 1.162a.217.217 0 01-.412 0l-.387-1.162A1.734 1.734 0 009.31 6.593l-1.162-.387a.217.217 0 010-.412l1.162-.387a1.734 1.734 0 001.097-1.097l.387-1.162zM13.863.099a.145.145 0 01.274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 010 .274l-.774.258a1.156 1.156 0 00-.732.732l-.258.774a.145.145 0 01-.274 0l-.258-.774a1.156 1.156 0 00-.732-.732l-.774-.258a.145.145 0 010-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"
          />
        </svg>
      </button>
    </nav>
    <label
      class="css-41wco4"
      height="7"
      width="20"
    >
      <span
        class="css-1grri41"
      />
      <div
        class="css-11gjszk"
      >
        <input
          class="css-1h6wv30"
          placeholder="Search"
          type="search"
          value=""
        />
        <svg
          aria-label="Search"
          class="css-5n1pld"
          fill="currentColor"
          height="2rem"
          role="img"
          viewBox="0 0 16 16"
          width="2rem"
        >
          <path
            d="M11.742 10.344a6.5 6.5 0 10-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 001.415-1.414l-3.85-3.85a1.007 1.007 0 00-.115-.1zM12 6.5a5.5 5.5 0 11-11 0 5.5 5.5 0 0111 0z"
          />
        </svg>
      </div>
      <span
        class="css-1reg724"
        role="alert"
      />
    </label>
    <main
      class="css-tee3o6"
    >
      <h1>
        Main article area
      </h1>
      <p>
        In this layout, we display the areas in source order for any screen less than 500 pixels wide. We go to a two column layout, and then to a three column layout by redefining the grid,and the placement of items on the grid.
      </p>
    </main>
    <footer
      class="css-1oqn78u"
    >
      <p>
        © 2022 
        <a
          class="css-1nz5qw1"
          href="https://github.com/esteban-90"
          target="_blank"
        >
          Esteban V.M.
        </a>
         
      </p>
      <p>
        All rights reserved
      </p>
    </footer>
  </div>
</DocumentFragment>
`;
